AsyncGenerator/AsyncIterable do not work with overloaded async def

**Bug Report**

When using `typing.overload` with an async function that returns an `AsyncGenerator` or `AsyncIterable`, the overloads fail with `Overloaded function implementation cannot produce return type of signature #` when defined using `async def`

**To Reproduce**

Failure case: https://mypy-play.net/?mypy=latest&python=3.11&gist=f9b64edbf09acd4b1952f5753669c2e7

```python
from typing import AsyncGenerator, overload

@overload
async def foo(*, a: int) -> AsyncGenerator[int, None]:
    ...
    
@overload
async def foo(*, b: int) -> AsyncGenerator[int, None]:
    ...

async def foo(*, a: int | None = None, b: int | None = None) -> AsyncGenerator[int, None]:
    if a is not None:
        for i in range(a):
            yield i
    elif b is not None:
        for i in range(b):
            yield i
    else:
        yield -1
```

Unexpected success case: https://mypy-play.net/?mypy=latest&python=3.11&gist=2fcc9c2372664f4ce257a9cec2849698

```python
from typing import AsyncGenerator, overload

@overload
def foo(*, a: int) -> AsyncGenerator[int, None]:
    ...
    
@overload
def foo(*, b: int) -> AsyncGenerator[int, None]:
    ...

async def foo(*, a: int | None = None, b: int | None = None) -> AsyncGenerator[int, None]:
    if a is not None:
        for i in range(a):
            yield i
    elif b is not None:
        for i in range(b):
            yield i
    else:
        yield -1
```

**Expected Behavior**

no errors when overloads are also defined with `async def` and error when overloads are defined with `def`

**Actual Behavior**

when overloads are defined with `async def`:

```
main.py:11: error: Overloaded function implementation cannot produce return type of signature 1  [misc]
main.py:11: error: Overloaded function implementation cannot produce return type of signature 2  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

when overloads are defined with `def`: no errors

**Your Environment**

- Mypy version used: 1.1.1 (mypy-play)
- Python version used: 3.11 (mypy-play)

--- a/docs/source/more_types.rst
+++ b/docs/source/more_types.rst
@@ -941,11 +941,11 @@ One common confusion is that the presence of a ``yield`` statement in an
        await arange(5)  # Error: Incompatible types in "await" (actual type "AsyncIterator[int]", expected type "Awaitable[Any]")
        reveal_type(await coroutine(5))  # Revealed type is "typing.AsyncIterator[builtins.int]"
 
-This can sometimes come up when trying to define base classes or Protocols:
+This can sometimes come up when trying to define base classes, Protocols or overloads:
 
 .. code-block:: python
 
-    from typing import AsyncIterator, Protocol
+    from typing import AsyncIterator, Protocol, overload
 
     class LauncherIncorrect(Protocol):
         # Because launch does not have yield, this has type
@@ -964,3 +964,17 @@ This can sometimes come up when trying to define base classes or Protocols:
             raise NotImplementedError
             if False:
                 yield 0
+
+    # The type of the overloads is independent of the implementation.
+    # In particular, their type is not affected by whether or not the
+    # implementation contains a `yield`.
+    # Use of `def`` makes it clear the type is Callable[..., AsyncIterator[int]],
+    # whereas with `async def` it would be Callable[..., Coroutine[Any, Any, AsyncIterator[int]]]
+    @overload
+    def launch(*, count: int = ...) -> AsyncIterator[int]: ...
+    @overload
+    def launch(*, time: float = ...) -> AsyncIterator[int]: ...
+
+    async def launch(*, count: int = 0, time: float = 0) -> AsyncIterator[int]:
+        # The implementation of launch is an async generator and contains a yield
+        yield 0

